File: test/proto/cpp/smoke.cpp
// Smoke test for the C++ test-provider port.
//
// Loads build/test/test.json (resolved relative to CWD; run from repo root),
// prints the functions list, total entries, expect-kind counts, input-kind
// counts, and the first getpath/basic entry.
//
//   g++ -std=c++17 -O2 test/proto/cpp/smoke.cpp -o /tmp/cpp_smoke
//   (cd /home/user/struct && /tmp/cpp_smoke)

#include <iostream>
#include <string>

#include "provider.hpp"

using namespace voxgig::testproto;

static std::string expectKindName(ExpectKind k) {
  switch (k) {
    case ExpectKind::VALUE:
      return "value";
    case ExpectKind::ERROR:
      return "error";
    case ExpectKind::MATCH:
      return "match";
    case ExpectKind::ABSENT:
      return "absent";
  }
  return "?";
}

static std::string inputKindName(InputKind k) {
  switch (k) {
    case InputKind::IN:
      return "in";
    case InputKind::ARGS:
      return "args";
    case InputKind::CTX:
      return "ctx";
  }
  return "?";
}

int main(int argc, char** argv) {
  std::string path = (argc > 1) ? argv[1] : "";
  TestProvider provider = TestProvider::load(path);

  // functions list
  std::vector<std::string> fns = provider.functions();
  std::cout << "functions: ";
  for (size_t i = 0; i < fns.size(); i++) {
    if (i) std::cout << ", ";
    std::cout << fns[i];
  }
  std::cout << "\n";

  // tally over all entries
  size_t total = 0;
  size_t valueN = 0, absentN = 0, matchN = 0, errorN = 0;
  size_t inN = 0, argsN = 0, ctxN = 0;
  for (const auto& fn : fns) {
    for (const auto& e : provider.entries(fn)) {
      total++;
      switch (e.expect.kind) {
        case ExpectKind::VALUE:
          valueN++;
          break;
        case ExpectKind::ABSENT:
          absentN++;
          break;
        case ExpectKind::MATCH:
          matchN++;
          break;
        case ExpectKind::ERROR:
          errorN++;
          break;
      }
      switch (e.input.kind) {
        case InputKind::IN:
          inN++;
          break;
        case InputKind::ARGS:
          argsN++;
          break;
        case InputKind::CTX:
          ctxN++;
          break;
      }
    }
  }

  std::cout << "total entries: " << total << " ; expect kinds: value=" << valueN
            << ", absent=" << absentN << ", match=" << matchN << ", error=" << errorN
            << " ; input kinds: in=" << inN;
  if (argsN) std::cout << ", args=" << argsN;
  if (ctxN) std::cout << ", ctx=" << ctxN;
  std::cout << "\n";

  // getpath/basic[0]
  std::vector<Entry> gp = provider.entries("getpath", std::optional<std::string>("basic"));
  if (!gp.empty()) {
    const Entry& e = gp[0];
    std::cout << "getpath/basic[0]: id=" << (e.id ? *e.id : std::string("null"))
              << ", doc=" << (e.doc ? "true" : "false")
              << ", input.kind=" << inputKindName(e.input.kind)
              << ", expect.kind=" << expectKindName(e.expect.kind) << ", expect.value=";
    if (e.expect.value.has_value())
      std::cout << jsonify(*e.expect.value);
    else
      std::cout << "(none)";
    std::cout << "\n";
  }

  return 0;
}

File: test/proto/zig/smoke.zig
// Smoke test for the Zig test-provider port. Prints a small summary of the
// loaded corpus so the numbers can be eyeballed against the canonical target:
//
//   functions: minor, getpath, inject, merge, transform, walk, validate,
//              select, sentinels
//   total entries: 1325
//   expect kinds: value=1181, absent=84, match=1, error=59
//   input  kinds: in=1325
//   getpath/basic[0]: id=getpath/basic#deep, doc=true, input.kind=in,
//                     expect.kind=value, expect.value=42
//
// Run (Zig 0.13): from test/proto/zig/  ->  zig run smoke.zig
// (No build.zig needed; provider.zig is a plain stdlib-only module.)

const std = @import("std");
const provider = @import("provider.zig");

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const g0 = gpa.allocator();

    // Use an arena for all transient provider output (function/group/entry
    // slices), freed at the end.
    var arena = std.heap.ArenaAllocator.init(g0);
    defer arena.deinit();
    const a = arena.allocator();

    const stdout = std.io.getStdOut().writer();

    var tp = try provider.TestProvider.load(g0, null);
    defer tp.deinit();

    // ── functions list ──
    const fns = try tp.functions(a);
    try stdout.writeAll("functions: ");
    for (fns, 0..) |f, i| {
        if (i != 0) try stdout.writeAll(", ");
        try stdout.writeAll(f);
    }
    try stdout.writeAll("\n");

    // ── totals + expect/input kind counts ──
    var total: usize = 0;
    var ev: usize = 0; // value
    var ea: usize = 0; // absent
    var em: usize = 0; // match
    var ee: usize = 0; // error
    var ii: usize = 0; // input .in
    var ia: usize = 0; // input .args
    var ic: usize = 0; // input .ctx

    for (fns) |f| {
        const es = try tp.entries(a, f);
        total += es.len;
        for (es) |e| {
            switch (e.expect.kind) {
                .value => ev += 1,
                .absent => ea += 1,
                .match => em += 1,
                .error_ => ee += 1,
            }
            switch (e.input.kind) {
                .in => ii += 1,
                .args => ia += 1,
                .ctx => ic += 1,
            }
        }
    }

    try stdout.print("total entries: {d}\n", .{total});
    try stdout.print("expect kinds: value={d}, absent={d}, match={d}, error={d}\n", .{ ev, ea, em, ee });
    try stdout.print("input kinds: in={d}", .{ii});
    if (ia != 0) try stdout.print(", args={d}", .{ia});
    if (ic != 0) try stdout.print(", ctx={d}", .{ic});
    try stdout.writeAll("\n");

    // ── getpath/basic[0] spotlight ──
    const gb = try tp.entriesGroup(a, "getpath", "basic");
    if (gb.len > 0) {
        const e = gb[0];
        const id = e.id orelse "<null>";
        try stdout.print("getpath/basic[0]: id={s}, doc={}, input.kind={s}, expect.kind={s}, expect.value=", .{
            id,
            e.doc,
            @tagName(e.input.kind),
            expectKindName(e.expect.kind),
        });
        if (e.expect.value) |v| {
            try printValue(stdout, v);
        } else {
            try stdout.writeAll("<none>");
        }
        try stdout.writeAll("\n");
    } else {
        try stdout.writeAll("getpath/basic[0]: <no entries>\n");
    }
}

fn expectKindName(k: provider.ExpectKind) []const u8 {
    return switch (k) {
        .value => "value",
        .error_ => "error",
        .match => "match",
        .absent => "absent",
    };
}

fn printValue(writer: anytype, v: std.json.Value) !void {
    switch (v) {
        .integer => |n| try writer.print("{d}", .{n}),
        .float => |n| try writer.print("{d}", .{n}),
        .string => |s| try writer.writeAll(s),
        .bool => |b| try writer.print("{}", .{b}),
        .null => try writer.writeAll("null"),
        else => try std.json.stringify(v, .{}, writer),
    }
}
